riscv/mmu.cc

```diff
@@ -161,6 +161,94 @@ void mmu_t::load_slow_path(reg_t addr, reg_t len, uint8_t* bytes, uint32_t xlate
   }
 }
 
+void mmu_t::clean_inval(reg_t addr, bool clean, bool inval)
+{
+  reg_t paddr = addr;
+  bool page = false, guest = false, access = false;
+  bool has_gva;
+  reg_t gpa;
+  access_type type = LOAD;
+  bool fault = true;
+  try {
+    paddr = translate(addr, 1, LOAD, 0);
+    type = LOAD;
+    fault = false;
+  } catch (trap_load_access_fault& t) {
+    access = true;
+    has_gva = t.has_gva();
+  } catch (trap_load_page_fault& t) {
+    page = true;
+    has_gva = t.has_gva();
+  } catch (trap_load_guest_page_fault& t) {
+    guest = true;
+    gpa = t.get_tval2();
+  }
+
+  if (fault) {
+    try {
+      paddr = translate(addr, 1, STORE, 0);
+      type = STORE;
+      fault = false;
+    } catch (trap_store_access_fault& t) {
+      access = true;
+      has_gva = t.has_gva();
+    } catch (trap_store_page_fault& t) {
+      page = true;
+      has_gva = t.has_gva();
+    } catch (trap_store_guest_page_fault& t) {
+      guest = true;
+      gpa = t.get_tval2();
+    }
+  }
+
+  if (fault) {
+    try {
+      paddr = translate(addr, 1, FETCH, 0);
+      type = FETCH;
+    } catch (trap_instruction_access_fault& t) {
+      access = true;
+      has_gva = t.has_gva();
+    } catch (trap_instruction_page_fault& t) {
+      page = true;
+      has_gva = t.has_gva();
+    } catch (trap_instruction_guest_page_fault& t) {
+      guest = true;
+      gpa = t.get_tval2();
+    }
+  }
+
+  if (access) {
+    trap_store_access_fault(has_gva, addr, 0, 0);
+  } else if (guest) {
+    trap_store_guest_page_fault(addr, gpa, 0);
+  } else if (page) {
+    trap_store_page_fault(has_gva, addr, 0, 0);
+  } else {
+    if (auto host_addr = sim->addr_to_mem(paddr)) {
+      if (tracer.interested_in_range(paddr, paddr + PGSIZE, type))
+        tracer.clean_invalidate(paddr, blocksz, clean, inval);
+      else
+        refill_tlb(addr, paddr, host_addr, type);
+    } else {
+      throw trap_store_access_fault((proc) ? proc->state.v : false, addr, 0, 0);
+    }
+
+    if (!matched_trigger) {
+      matched_trigger = trigger_exception(type == LOAD? OPERATION_LOAD :
+                                          type == STORE? OPERATION_STORE : OPERATION_EXECUTE,
+                                          addr, 0);
+      if (matched_trigger)
+        throw *matched_trigger;
+    }
+  }
+}
```

Comment:
This seems excessive, especially all the special handling of exceptions.

Reply:
In the spec,  " A cache-block management instruction is permitted to access the specified cache block whenever a load
instruction, store instruction, or instruction fetch is permitted to access the corresponding physical addresses."   That means it may do LOAD STORE or FETCH check. However,  current translate progress can only pass one type at a time. So in above implementation it divide into three steps to check each access type: 

- No matter which access type pass the translation progress, cbo instruction will execute successfully. 
- If all of them trigger exceptions. it will combine the final exception result according which exception is triggered at last (birefly access  fault> guest page fault > page fault)

Reply:
Since stores don't succeed where loads fail, it should be possible to only check for load or fetch permissions.

Furthermore, the interaction between CBOs and triggers is implementation-defined, so we're free to skip that part.

We're also under no obligation to refill the TLB.  For these instructions, simple code is 

And finally, we don't need to be quite so precise about what `type` gets passed to `tracer.interested_in_range`.  Between all of those, I think this can be simplified substantially.  Something like this:

```
void mmu_t::clean_inval(reg_t addr, bool clean, bool inval)
{ 
  reg_t paddr;
  
  try {
    paddr = translate(addr, 1, LOAD, 0);
  } catch (...) {
    // Try again with fetch permissions 
    try {
      paddr = translate(addr, 1, FETCH, 0);
    } catch (trap_instruction_access_fault& t) {
      trap_store_access_fault(t.has_gva(), addr, 0, 0);
    } catch (trap_instruction_page_fault& t) {
      trap_store_page_fault(t.has_gva(), addr, 0, 0);
    } catch (trap_instruction_guest_page_fault& t) {
      trap_store_guest_page_fault(addr, t.get_tval2(), 0);
    } catch (...) {
      abort();
    }
  }

  if (auto host_addr = sim->addr_to_mem(paddr)) { 
    if (tracer.interested_in_range(paddr, paddr + PGSIZE, LOAD))
      tracer.clean_invalidate(paddr, blocksz, clean, inval);
  } else {
    throw trap_store_access_fault((proc) ? proc->state.v : false, addr, 0, 0);
  }
}
```

Reply:
Would it make any sense to modify `translate()` so that we can pass a list of access types? So we could do:
```
paddr = translate(addr, 1, LOAD | FETCH | STORE, 0);
```

Reply:
I'd need to see it both ways to judge.  I'm worried that doing so would make the already complex page-table-walking code messier.  This approach is certainly ugly, but at least it is contained.

Reply:
> Since stores don't succeed where loads fail, it should be possible to only check for load or fetch permissions.

The PMP check for LOAD/STORE seems different. So, I think we should add LOAD access here.

Reply:
> void mmu_t::clean_inval(reg_t addr, bool clean, bool inval)
> { 
>   reg_t paddr;
>   
>   try {
>     paddr = translate(addr, 1, LOAD, 0);
>   } catch (...) {
>     // Try again with fetch permissions 
>     try {
>       paddr = translate(addr, 1, FETCH, 0);
>     } catch (trap_instruction_access_fault& t) {
>       trap_store_access_fault(t.has_gva(), addr, 0, 0);
>     } catch (trap_instruction_page_fault& t) {
>       trap_store_page_fault(t.has_gva(), addr, 0, 0);
>     } catch (trap_instruction_guest_page_fault& t) {
>       trap_store_guest_page_fault(addr, t.get_tval2(), 0);
>     } catch (...) {
>       abort();
>     }
>   }

The final exception should combine the exception of diffrent access types. In this implementation the exception type of FETCH will be the final exception type.

Reply:
> Would it make any sense to modify `translate()` so that we can pass a list of access types? So we could do:
> 
> ```
> paddr = translate(addr, 1, LOAD | FETCH | STORE, 0);
> ```

There is two problem in this way:

- In the begin of translation before walk, the virt parameter should take the debug mode an  MPRV, MPP, MPV .etc into account for only LOAD and STORE. So how do we decide the virt parameter for LOAD | FETCH | STORE.
- The translation may  be devided into several levels. The access types may be reduced if some access types cannot pass the check of one specific level. We need a good way to transport which types cannot pass check between different levels  and then update the type for next level.

Reply:
OK, I concede. I agree it would get messy elsewhere.

Won't hardware need to do all that mess, though?

Reply:
Indeed.

Reply:
Sorry. I have no idea how hardware implement it in an easy way.

Reply:
Regardless, I'm still going to insist that checking for store permissions is unnecessary.  At the very least, we should be able to simplify it to something like the following:

```void mmu_t::clean_inval(reg_t addr, bool clean, bool inval)
{
  reg_t paddr = addr;
  bool page = false, guest = false, access = false;
  bool has_gva;
  reg_t gpa;

  bool fault = true;
  try {
    paddr = translate(addr, 1, LOAD, 0);
  } catch (trap_load_access_fault& t) {
    access = true;
    has_gva = t.has_gva();
  } catch (trap_load_page_fault& t) {
    page = true;
    has_gva = t.has_gva();
  } catch (trap_load_guest_page_fault& t) {
    guest = true;
    gpa = t.get_tval2();
  }

  if (fault) {
    try {
      paddr = translate(addr, 1, FETCH, 0);
    } catch (trap_instruction_access_fault& t) {
      access = true;
      has_gva = t.has_gva();
    } catch (trap_instruction_page_fault& t) {
      page = true;
      has_gva = t.has_gva();
    } catch (trap_instruction_guest_page_fault& t) {
      guest = true;
      gpa = t.get_tval2();
    }
  }

  if (access) {
    trap_store_access_fault(has_gva, addr, 0, 0);
  } else if (guest) {
    trap_store_guest_page_fault(addr, gpa, 0);
  } else if (page) {
    trap_store_page_fault(has_gva, addr, 0, 0);
  } else {
    if (auto host_addr = sim->addr_to_mem(paddr)) {
      if (tracer.interested_in_range(paddr, paddr + PGSIZE, LOAD))
        tracer.clean_invalidate(paddr, blocksz, clean, inval);
    } else {
      throw trap_store_access_fault((proc) ? proc->state.v : false, addr, 0, 0);
    }
  }
}
```

Reply:
> This seems excessive, especially all the special handling of exceptions.

Maybe we should take all of LOAD, STORE and FETCH into consideration since both icache and dcache need to clean or invalidate.  
```
       if (tracer.interested_in_range(paddr, paddr + PGSIZE, LOAD) ||
          tracer.interested_in_range(paddr, paddr + PGSIZE, FETCH))
          tracer.clean_invalidate(paddr, blocksz, clean, inval);
```

Reply:
I don't think so, because these CBOs don't have any effect on an incoherent I$.  Simply checking for loads (so that the D$ model sees the invalidate) should suffice.

Reply:
I wonder, If only D$ model, why we check FETCH access.

Reply:
OK. Checking LOAD is suffiice in current implementation. It will return true if one of cache return true.

Reply:
However if we only enable IC and L2, then this will never clean and invalidate caches. 
Maybe I have gone into the wrong logic. The question why we shouldn't have effect on IC?

Reply:
> Regardless, I'm still going to insist that checking for store permissions is unnecessary. At the very least, we should be able to simplify it to something like the following:
> 
> ```
> void mmu_t::clean_inval(reg_t addr, bool clean, bool inval)
> {
>   reg_t paddr = addr;
>   bool page = false, guest = false, access = false;
>   bool has_gva;
>   reg_t gpa;
> 
>   bool fault = true;
>   try {
>     paddr = translate(addr, 1, LOAD, 0);
>   } catch (trap_load_access_fault& t) {
>     access = true;
>     has_gva = t.has_gva();
>   } catch (trap_load_page_fault& t) {
>     page = true;
>     has_gva = t.has_gva();
>   } catch (trap_load_guest_page_fault& t) {
>     guest = true;
>     gpa = t.get_tval2();
>   }
> 
>   if (fault) {
>     try {
>       paddr = translate(addr, 1, FETCH, 0);
>     } catch (trap_instruction_access_fault& t) {
>       access = true;
>       has_gva = t.has_gva();
>     } catch (trap_instruction_page_fault& t) {
>       page = true;
>       has_gva = t.has_gva();
>     } catch (trap_instruction_guest_page_fault& t) {
>       guest = true;
>       gpa = t.get_tval2();
>     }
>   }
> 
>   if (access) {
>     trap_store_access_fault(has_gva, addr, 0, 0);
>   } else if (guest) {
>     trap_store_guest_page_fault(addr, gpa, 0);
>   } else if (page) {
>     trap_store_page_fault(has_gva, addr, 0, 0);
>   } else {
>     if (auto host_addr = sim->addr_to_mem(paddr)) {
>       if (tracer.interested_in_range(paddr, paddr + PGSIZE, LOAD))
>         tracer.clean_invalidate(paddr, blocksz, clean, inval);
>     } else {
>       throw trap_store_access_fault((proc) ? proc->state.v : false, addr, 0, 0);
>     }
>   }
> }
> ```
OK. I'll discuss this with David again.

Reply:
A few thoughts (hopefully not too out of context at this point):
- A load access should cover a store access, so it should be sufficient to check just for load (I guided Weiwei to add store to the list of permissions checks since the spec calls store accesses out)
- The spec explicitly adds "store access" because several people balked at the explanation that load access was sufficient
- There is a difference with PMP and PMA permissions since W access is allowed without R access, but the current spec says this combination is UNSPECIFIED for the purposes of CBOs
- Non-coherent I-caches are not affected because such caches are not directly or indirectly accessed by loads or stores; a coherent I-cache, on the other hand, can be directly or indirectly accessed (in the latter case via snoops/probes/etc.) so the CBOs must operate on coherent I-caches (the operation is required because an instruction fetch may bring the block into an "I-cache" but that same cache, in some implementations, could later provide the data on a load or store access)
- The previous bullet explains why instruction fetch is present in the list of accesses (though my suspicion is that, for a number of reasons, we may need to adjust this definition when we finally define ops for I/D coherence)

Looking at how implementations might approach these checks, I note that translation is serialized with PMP and PMA checks, so the basic thought is that an implementation performs translation, subject to MPP, MPRV, and friends, and remembers the types of accesses allowed by the translation process (or takes a translation fault, of course). Each of those allowed access types is compared against the permissions granted by the PMP and PMAs, and if any are left standing, the CBO is allowed. Depending on where you start with an implementation for basic address translation and access permission checks, this process may be easy or hard. (I was hoping that it would not be hard.....)

Reply:
Thanks for your explanation. I'll fix the current implementation firstly. And Then try the way to check LOAD/STORE/FETCH.

Reply:
> * In the begin of translation before walk, the virt parameter should take the debug mode an  MPRV, MPP, MPV .etc into account for only LOAD and STORE. So how do we decide the virt parameter for LOAD | FETCH | STORE.
> 
> * The translation may  be devided into several levels. The access types may be reduced if some access types cannot pass the check of one specific level. We need a good way to transport which types cannot pass check between different levels  and then update the type for next level.

@dkruckemyer-ventana the second problem may be not difficult to solve since only the last level of translation is related to actual access type. However, the translation progress is totally different(including pte address and content for each level of translation) if virt is different for FETCH and LOAD/STORE according debug mode  and MPRV, MPP, MPV .etc.

And this may introduce  an additional problem for current implementation. In my understanding, our combination of exceptions is based on the assumption that all the exception triggered before the last level of G-stage translation is the same (access fault, page fault, or guest page fault) for different access type(the part of progress is unrelated to truly access type).  In the last level of G-stage translation, they may trigger different fault, however the trigger sequence is  check for page fault, then guest page fault, and finally access fault (by PMP check) . So the priority is access > guest page fault > page fault.

Now, the first problem may make the pte address and content for each level of translation totally different  for FETCH and LOAD/STORE. This will break above base assumption. Do I misunderstand anything?

Reply:
I’m not sure what the confusion is - my proposed solution still appears to be correct.

Reply:
The problem is the priority of exceptions, which implement as follows:
```
  if (access) {
    trap_store_access_fault(has_gva, addr, 0, 0);
  } else if (guest) {
    trap_store_guest_page_fault(addr, gpa, 0);
  } else if (page) {
    trap_store_page_fault(has_gva, addr, 0, 0);
  } 
```

Reply:
You had me stumped there for a moment, but I think everything is OK in the spec....

The translation is based on the effective address of the CBO, which, like other explicit load and store accesses, is affected by MPRV. Once the translation is performed, the permissions are checked against the RWX bits in the various mechanisms. So the translation and permission check is *not* performed as if the access were an instruction fetch.

But this likely highlights an issue with performing a FETCH translation by itself.... (I was wrong in thinking that was the case.)

Reply:
Perhaps a way to approach this is to perform two LOAD accesses, one with MXR=0 and one with MXR=1? (Though I haven't quite thought through whether that explodes with virtualization....)

Reply:
MXR doesn't affect PMP, and so that wouldn't suffice to allow CBOs to regions that have a R=0 X=1 PMP configuration.

Reply:
I have modified translation progress to support check for cbo ops If virt is not a problem.